stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <syncstream> is not supported when compiling with /clr:pure
+#endif // _M_CEE_PURE
+
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+struct _Mutex_count_pair {
+    shared_mutex _Mutex;
+    uint64_t _Ref_count = 0;
+};
+
+_CRTIMP2_PURE map<void*, _Mutex_count_pair>& __cdecl _Get_mutex_map();

```

Comment:
Functions that need to worry about `/clr:pure` are typically `_CRTIMP2_PURE ReturnType __CLRCALL_PURE_OR_CDECL`. I see that some of our multithreading functions are `_CRTIMP2_PURE ReturnType __cdecl` but this is probably a (cosmetic) mistake as the multithreading code typically forbids `/clr:pure`; they should probably be `_CRTIMP2 ReturnType __cdecl`. Our new code prefers flat C exports with `__stdcall` (I believe because it's marginally more efficient); `xfilesystem_abi.h` and `stl/src/atomic_wait.cpp` are examples.

However, before we get to the question of exporting and calling conventions, there's the ABI question. For somewhat arcane binary compatibility reasons, we currently avoid *adding* exports to the STL's DLL (much less modifying/removing exports which outright breaks bincompat). When we need to add separately compiled machinery, our first preference is to "inject" it into the DLL's import LIB, which effectively statically links it; this is the approach we took with Filesystem. When we additionally need global data (as is the case for memory resources, codecvt IDs, and atomic waits, as I understand it), the import LIB approach isn't sufficient, and we need to go further and add a "satellite DLL" - most recently done for atomic waits.

If we need global mutexes (I am not yet up to speed on syncstream implementation details), then the satellite DLL approach may be necessary. This is a surprisingly enormous headache (beyond just the build system changes for GitHub, there are MSVC-internal build changes *and* there's a bunch of poorly understood MSVC-internal changes that need to happen) but there is no other choice than waiting for vNext which we have tried to avoid for new features (successfully so far).

Reply:
@StephanTLavavej Can it be added to atomic_wait DLL since c++latest has unstable ABI?

Reply:
As for now, I added the source file to build with the `IMPLIB_SOURCES` group, which I suppose is the static import library.

I am experimenting with building it with other targets, but it fails to link with `shared_mutex::lock` and `shared_mutex::unlock`. I'll let you decide the next steps since I have little experience with STL and CRT interactions.

Reply:
That's an excellent idea, @AdamBucior. Aside from making the name incomplete, I see no reason why we couldn't do that, and the name being incompletely descriptive is much less of a headache for us *and* users than adding a new satellite.

Reply:
C+17 parallel algorithms rely on atomic wait DLL.
I still see no issues as any version of atomic wait dll would have support for parallel algorithms.